INSTALL.md: cygwin details and add source code link (#17485)
curl/curl: docs/INSTALL.md

Lots of people download binary distributions of curl and libcurl. This
document does not describe how to install curl or libcurl using such a binary
package. This document describes how to compile, build and install curl and

> libcurl from [source code](https://curl.se/download.html#:~:text=offer%2E-,Packages).

Reviewer: ```suggestion
libcurl from [source code](https://curl.se/download.html).
```
Drop anchor.

in your path that would conflict with `cygwin` commands, making sure you have the `sh`
executable in `/bin/` or you see the configure fail toward the end.

Download the setup installer from
[`cygwin`](https://cygwin.com/) to begin. Additional `cygwin`
packages are needed for the install. For more on installing packages visit
[`cygwin setup`](https://www.cygwin.com/faq/faq.html#faq.setup.cli).

Either run setup-x86_64.exe, then search and select packages individually, or try:

    setup-x86_64.exe -P binutils, gcc-core, libpsl-devel, libtool, perl, make

If the latter, matching packages should appear in the install rows (*is fickle though*) after selecting
the download site i.e. `https://mirrors.kernel.org`. In either case, follow the GUI prompts
until you reach the "Select Packages" window; then select packages, click next, and finish
the `cygwin` package installation.

Download the latest version of the `cygwin` packages required (*and suggested*) for a successful install:

<details>
    <summary>Package List</summary>

```
 binutil - required
 gcc-core - required
 libpsl-devel - required
 libtool - required
 perl - required
 make - required

>  - NOTE - if there is an error regarding cmake, open the cygwin terminal, and run:

Reviewer: ```suggestion
 - NOTE - if there is an error regarding make, open the cygwin terminal, and run:
```
Is cmake correct here?
In what case can this error be popping up?

Author: Thank you. I missed that.

   ln -s /usr/bin/make /usr/bin/gmake
 libuv1 - suggested
 libarchive13 - suggested

>  libcurl4 - suggested

Reviewer: Perhaps drop these also?
libarchive is unrelated to curl, libcurl4 is what we build here, libuv1 is an option but a niche one not used by these instructions (it's also slow when tested with curl on Cygwin)

## Cygwin

[curl source code](https://curl.se/download.html#Win64cygwin)

Almost identical to the Unix installation. Essentially run the configure script in the
curl source tree root with `sh configure`, then run `make`.

To expand on building with `cygwin` first ensure it is in your path, and there are no
conflicting tools (*i.e. Chocolatey with sed package*). If so move `cygwin` ahead of any items
in your path that would conflict with `cygwin` commands, making sure you have the `sh`
executable in `/bin/` or you see the configure fail toward the end.

Download the setup installer from

> [`cygwin`](https://cygwin.com/#:~:text=Installing%20Cygwin) to begin. Additional `cygwin`

Reviewer: ```suggestion
[`cygwin`](https://cygwin.com/) to begin. Additional `cygwin`
```
Drop weird anchor?

packages are needed for the install. For more on installing packages visit
[`cygwin setup`](https://www.cygwin.com/faq/faq.html#faq.setup.cli).

Either run setup-x86_64.exe, then search and select packages individually, or try:

    setup-x86_64.exe -P binutil gcc-core libpsl-devel libtool perl make

If the latter, matching packages should appear in the install rows after selecting
the download site i.e. `https://mirrors.kernel.org`. In either case, follow the GUI prompts
until you reach the "Select Packages" window; then select packages, click next, and finish
the `cygwin` package installation.

Unless specified, download the latest version of the package. The `cygwin` packages required (*and suggested*) for a successful install are:

<details>
    <summary>Package List</summary>

```
 binutil - required
 gcc-core - required
 libpsl-devel - required
 libtool (latest or 2.5.4-1) - required
 perl - required
 make - required
 - NOTE - if there is an error regarding cmake, open the cygwin terminal, and run:
   ln -s /usr/bin/make /usr/bin/gmake
 libuv1 - suggested
 libarchive13 - suggested
 libcurl4 - suggested
 libexpat1 - suggested
 libjsoncpp25 - suggested
 librhash0 - suggested

> ```

Reviewer: This information is prone to go stale by time. Are we sure we to keep this updated, is it adding useful information for Cygwin users?

Author: For the required packages - yes, I think so. For the suggested ones - maybe. But removed libexpat1, libjsoncpp25, and librhash0 from the list; and after re-installing using `sh configure --without-ssl`, the installs were consistently successful. When using `sh configure --with-openssl`, the configuration works, but when `make` runs; I get the most consistent build error:

    make[2]: Entering directory '/cygdrive/c/Users/WDAGUtilityAccount/Desktop/sandbox/curl/lib'
      CC       libcurl_la-curl_ntlm_core.lo
      CCLD     libcurl.la
    ../libtool: eval: line 1890: syntax error near unexpected token `|'
    ../libtool: eval: line 1890: `/usr/bin/nm -B  .libs/libcurl_la-altsvc.o .libs/libcurl_la-amigaos.o .libs/libcurl_la-asyn-ares.o .libs/libcurl_la-asyn-thread.o .libs/libcurl_la-base64.o .libs/libcurl_la-bufq.o .libs/libcurl_la-bufref.o .libs/libcurl_la-cf-h1-proxy.o .libs/libcurl_la-cf-h2-proxy.o .libs/libcurl_la-cf-haproxy.o .libs/libcurl_la-cf-https-connect.o .libs/libcurl_la-cf-socket.o .libs/libcurl_la-cfilters.o .libs/libcurl_la-conncache.o .libs/libcurl_la-cshutdn.o .libs/libcurl_la-connect.o .libs/libcurl_la-content_encoding.o .libs/libcurl_la-cookie.o .libs/libcurl_la-curl_addrinfo.o .libs/libcurl_la-curl_des.o .libs/libcurl_la-curl_endian.o .libs/libcurl_la-curl_fnmatch.o .libs/libcurl_la-curl_get_line.o .libs/libcurl_la-curl_gethostname.o .libs/libcurl_la-curl_gssapi.o .libs/libcurl_la-curl_memrchr.o .libs/libcurl_la-curl_multibyte.o .libs/libcurl_la-curl_ntlm_core.o .libs/libcurl_la-curl_range.o .libs/libcurl_la-curl_rtmp.o .libs/libcurl_la-curl_sasl.o .libs/libcurl_la-curl_sha512_256.o .libs/libcurl_la-curl_sspi.o .libs/libcurl_la-curl_threads.o .libs/libcurl_la-curl_trc.o .libs/libcurl_la-cw-out.o .libs/libcurl_la-cw-pause.o .libs/libcurl_la-dict.o .libs/libcurl_la-doh.o .libs/libcurl_la-dynbuf.o .libs/libcurl_la-dynhds.o .libs/libcurl_la-easy.o .libs/libcurl_la-easygetopt.o .libs/libcurl_la-easyoptions.o .libs/libcurl_la-escape.o .libs/libcurl_la-file.o .libs/libcurl_la-fileinfo.o .libs/libcurl_la-fopen.o .libs/libcurl_la-formdata.o .libs/libcurl_la-ftp.o .libs/libcurl_la-ftplistparser.o .libs/libcurl_la-getenv.o .libs/libcurl_la-getinfo.o .libs/libcurl_la-gopher.o .libs/libcurl_la-hash.o .libs/libcurl_la-hash_offt.o .libs/libcurl_la-headers.o .libs/libcurl_la-hmac.o .libs/libcurl_la-hostasyn.o .libs/libcurl_la-hostip.o .libs/libcurl_la-hostip4.o .libs/libcurl_la-hostip6.o .libs/libcurl_la-hostsyn.o .libs/libcurl_la-hsts.o .libs/libcurl_la-http.o .libs/libcurl_la-http1.o .libs/libcurl_la-http2.o .libs/libcurl_la-http_aws_sigv4.o .libs/libcurl_la-http_chunks.o .libs/libcurl_la-http_digest.o .libs/libcurl_la-http_negotiate.o .libs/libcurl_la-http_ntlm.o .libs/libcurl_la-http_proxy.o .libs/libcurl_la-httpsrr.o .libs/libcurl_la-idn.o .libs/libcurl_la-if2ip.o .libs/libcurl_la-imap.o .libs/libcurl_la-inet_ntop.o .libs/libcurl_la-inet_pton.o .libs/libcurl_la-krb5.o .libs/libcurl_la-ldap.o .libs/libcurl_la-llist.o .libs/libcurl_la-macos.o .libs/libcurl_la-md4.o .libs/libcurl_la-md5.o .libs/libcurl_la-memdebug.o .libs/libcurl_la-mime.o .libs/libcurl_la-mprintf.o .libs/libcurl_la-mqtt.o .libs/libcurl_la-multi.o .libs/libcurl_la-multi_ev.o .libs/libcurl_la-netrc.o .libs/libcurl_la-nonblock.o .libs/libcurl_la-noproxy.o .libs/libcurl_la-openldap.o .libs/libcurl_la-parsedate.o .libs/libcurl_la-pingpong.o .libs/libcurl_la-pop3.o .libs/libcurl_la-progress.o .libs/libcurl_la-psl.o .libs/libcurl_la-rand.o .libs/libcurl_la-rename.o .libs/libcurl_la-request.o .libs/libcurl_la-rtsp.o .libs/libcurl_la-select.o .libs/libcurl_la-sendf.o .libs/libcurl_la-setopt.o .libs/libcurl_la-sha256.o .libs/libcurl_la-share.o .libs/libcurl_la-slist.o .libs/libcurl_la-smb.o .libs/libcurl_la-smtp.o .libs/libcurl_la-socketpair.o .libs/libcurl_la-socks.o .libs/libcurl_la-socks_gssapi.o .libs/libcurl_la-socks_sspi.o .libs/libcurl_la-speedcheck.o .libs/libcurl_la-splay.o .libs/libcurl_la-strcase.o .libs/libcurl_la-strdup.o .libs/libcurl_la-strequal.o .libs/libcurl_la-strerror.o .libs/libcurl_la-strparse.o .libs/libcurl_la-system_win32.o .libs/libcurl_la-telnet.o .libs/libcurl_la-tftp.o .libs/libcurl_la-timediff.o .libs/libcurl_la-timeval.o .libs/libcurl_la-transfer.o .libs/libcurl_la-url.o .libs/libcurl_la-urlapi.o .libs/libcurl_la-version.o .libs/libcurl_la-version_win32.o .libs/libcurl_la-warnless.o .libs/libcurl_la-ws.o vauth/.libs/libcurl_la-cleartext.o vauth/.libs/libcurl_la-cram.o vauth/.libs/libcurl_la-digest.o vauth/.libs/libcurl_la-digest_sspi.o vauth/.libs/libcurl_la-gsasl.o vauth/.libs/libcurl_la-krb5_gssapi.o vauth/.libs/libcurl_la-krb5_sspi.o vauth/.libs/libcurl_la-ntlm.o vauth/.libs/libcurl_la-ntlm_sspi.o vauth/.libs/libcurl_la-oauth2.o vauth/.libs/libcurl_la-spnego_gssapi.o vauth/.libs/libcurl_la-spnego_sspi.o vauth/.libs/libcurl_la-vauth.o vtls/.libs/libcurl_la-bearssl.o vtls/.libs/libcurl_la-cipher_suite.o vtls/.libs/libcurl_la-gtls.o vtls/.libs/libcurl_la-hostcheck.o vtls/.libs/libcurl_la-keylog.o vtls/.libs/libcurl_la-mbedtls.o vtls/.libs/libcurl_la-mbedtls_threadlock.o vtls/.libs/libcurl_la-openssl.o vtls/.libs/libcurl_la-rustls.o vtls/.libs/libcurl_la-schannel.o vtls/.libs/libcurl_la-schannel_verify.o vtls/.libs/libcurl_la-sectransp.o vtls/.libs/libcurl_la-vtls.o vtls/.libs/libcurl_la-vtls_scache.o vtls/.libs/libcurl_la-vtls_spack.o vtls/.libs/libcurl_la-wolfssl.o vtls/.libs/libcurl_la-x509asn1.o vquic/.libs/libcurl_la-curl_msh3.o vquic/.libs/libcurl_la-curl_ngtcp2.o vquic/.libs/libcurl_la-curl_osslq.o vquic/.libs/libcurl_la-curl_quiche.o vquic/.libs/libcurl_la-vquic.o vquic/.libs/libcurl_la-vquic-tls.o vssh/.libs/libcurl_la-libssh.o vssh/.libs/libcurl_la-libssh2.o vssh/.libs/libcurl_la-curl_path.o vssh/.libs/libcurl_la-wolfssh.o .libs/libcurl_la-dllmain.o   |  | /usr/bin/sed -e '/^[BCDGRS][ ]/s/.*[ ]\([^ ]*\)/\1 DATA/;s/^.*[ ]__nm__\([^ ]*\)[ ][^ ]*/\1 DATA/;/^I[ ]/d;/^[AITW][ ]/s/.* //' | sort | uniq > .libs/libcurl.exp'
    make[2]: *** [Makefile:1852: libcurl.la] Error 2
    make[2]: Leaving directory '/cygdrive/c/Users/WDAGUtilityAccount/Desktop/sandbox/curl/lib'
    make[1]: *** [Makefile:1649: all] Error 2
    make[1]: Leaving directory '/cygdrive/c/Users/WDAGUtilityAccount/Desktop/sandbox/curl/lib'
    make: *** [Makefile:628: all-recursive] Error 1

Reviewer: Is there a way to replicate this e.g. in CI? Such build has been running in CI for a long time without bumping into this issue with TLS enabled.

</details>

Once all the packages have been installed, begin the process of installing curl from the source code:

 <details>
     <summary>configure_options</summary>

```

>     --with-amissl
>     --with-bearssl
>     --with-gnutls
>     --with-mbedtls
>     --with-openssl (also works for BoringSSL and LibreSSL)
>     --with-rustls
>     --with-schannel
>     --with-secure-transport
>     --with-wolfssl
>     --without-ssl

Reviewer: ```suggestion
    --with-gnutls
    --with-mbedtls
    --with-openssl (also works for OpenSSL forks)
    --with-rustls
    --with-wolfssl
    --without-ssl
```
Drop TLS backends that definitely don't work on Cygwin.
Does Rustls support Cygwin?

`--with-openssl` supports any OpenSSL fork, including AWS-LC, quictls.

Author: I have not tried using `sh configure --with-rustls` in the install process (*see [support repo](https://github.com/jhauga/curl-cygwin-install)*), and got an error during the configuration.

```

 </details>

>  1. ` sh configure <configure_options> `
>  2. ` make `

Reviewer: ```suggestion
 1. `sh configure <configure_options>`
 2. `make`
```

## AmigaOS

> [curl source code](https://curl.se/download.html#AmigaOS)

Reviewer: All platforms do build from the same source tarball. It seems overkill to link to it for each individual platform. These links also point to distro endpoints, most of the time binaires, not the to the curl sources.

(this applies to all sections, with `curl source code` links.)
